File: LC_57.java
// 57. Insert Interval

class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        int newStart = newInterval[0];
        int newEnd = newInterval[1];
        List<int[]> result = new ArrayList<>();
        
        for (int[] interval : intervals) {
            int curStart = interval[0];
            int curEnd = interval[1];
            if (curEnd < newStart)
                result.add(interval);
            else if (newEnd < curStart) {
                result.add(new int[] {newStart, newEnd});
                newStart = curStart;
                newEnd = curEnd;
            } else {
                newStart = Math.min(curStart, newStart);
                newEnd = Math.max(curEnd, newEnd);
            }
        }
        result.add(new int[] {newStart, newEnd});
        
        return result.toArray(new int[result.size()][2]);
    }
}

File: LC_57.py
# 57. Insert Interval

class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        result = []
        prev = newInterval
        
        for cur in intervals:
            # if current is completely past previous interval
            if prev[1] < cur[0]:
                result.append(prev)
                prev = cur
            # if current is completely before previous interval
            elif cur[1] < prev[0]:
                result.append(cur)
            # if there's an overlap
            else:
                prev[0] = min(prev[0], cur[0])
                prev[1] = max(prev[1], cur[1])
        
        result.append(prev)
        
        return result
